relay-server/src/endpoints/project_configs.rs

```diff
@@ -1,39 +1,71 @@
 use std::collections::HashMap;
+use std::sync::Arc;
 
 use actix::prelude::*;
-use actix_web::{Error, Json};
+use actix_web::{Error, FromRequest, Json};
 use futures::{future, Future};
 use serde::{Deserialize, Serialize};
 
 use relay_common::ProjectKey;
 
 use crate::actors::project::{LimitedProjectState, ProjectState};
-use crate::actors::project_cache::{GetProjectState, ProjectCache};
+use crate::actors::project_cache::{GetCachedProjectState, GetProjectState, ProjectCache};
 use crate::extractors::SignedJson;
 use crate::service::ServiceApp;
 use crate::utils::ErrorBoundary;
 
+/// V2 version of this endpoint.
+///
+/// The request is a list of [`ProjectKey`]s and the response a list of [`ProjectStateWrapper`]s
+const ENDPOINT_V2: u16 = 2;
+
+/// V3 version of this endpoint.
+///
+/// This version allows returning `pending` project configs, so anything in the cache can be
+/// returned directly.  The pending projectconfigs will also be fetched from upstream, so
+/// next time a downstream relay polls for this it is hopefully in our cache and will be
+/// returned, or a further poll ensues.
+const ENDPOINT_V3: u16 = 3;
+
 /// Helper to deserialize the `version` query parameter.
-#[derive(Debug, Deserialize)]
+#[derive(Clone, Copy, Debug, Deserialize)]
 struct VersionQuery {
     version: u16,
 }
 
+impl VersionQuery {
+    fn from_request(req: &actix_web::Request) -> Self {
+        let query = req.uri().query().unwrap_or("");
+        serde_urlencoded::from_str::<VersionQuery>(query).unwrap_or(VersionQuery { version: 0 })
+    }
+}
+
+impl<S> FromRequest<S> for VersionQuery {
+    type Config = ();
+    type Result = Self;
+
+    fn from_request(req: &actix_web::HttpRequest<S>, _: &Self::Config) -> Self::Result {
+        Self::from_request(&*req)
+    }
+}
+
 /// Checks for a specific `version` query parameter.
-struct VersionPredicate(u16);
+struct VersionPredicate;
 
 impl<S> actix_web::pred::Predicate<S> for VersionPredicate {
     fn check(&self, req: &actix_web::Request, _: &S) -> bool {
-        let query = req.uri().query().unwrap_or("");
-        serde_urlencoded::from_str::<VersionQuery>(query)
-            .map(|query| query.version == self.0)
-            .unwrap_or(false)
+        let query = VersionQuery::from_request(req);
+        query.version >= ENDPOINT_V2 && query.version <= ENDPOINT_V3
```

Comment:
the previous move from v1 to v2 has been implemented such that relay only supports one version, forwards all other requests at the HTTP layer, and all legacy compat stuff is handled in sentry. did we decide against continuing this path because we fear the increased traffic on python web workers coming from existing v2 customer relays? that's a bit unfortunate, it makes versioning more complicated. I think sentry will have to support all versions forever anyway. I thought it would be possible to weather through the increased traffic on python web workers until customers upgrade their relays.

Reply:
The basic model is still the same. Relay supports V2 and V3 now, handling all deployed versions of Relay without relying on the HTTP endpoint in Sentry. It will forward V4+ though, just like we used to forward future versions before. So the difference now is just that we also keep supporting the "old" V2 for performance reasons.


> that's a bit unfortunate, it makes versioning more complicated.

Agreed, although we can drop V2 support some time in the future, returning to supporting just a single version. Given the current implementation, do you think this is too much added complexity in this case? Personally, to me this seems still quite low-overhead.